File: LeetCode/Explore/Queue & Stack/284. Peeking Iterator.py
# Below is the interface for Iterator, which is already defined for you.
#
# class Iterator:
#     def __init__(self, nums):
#         """
#         Initializes an iterator object to the beginning of a list.
#         :type nums: List[int]
#         """
#
#     def hasNext(self):
#         """
#         Returns true if the iteration has more elements.
#         :rtype: bool
#         """
#
#     def next(self):
#         """
#         Returns the next element in the iteration.
#         :rtype: int
#         """

class PeekingIterator:
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self._iterator = iterator
        self._peeked_value = None
        

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        if self._peeked_value == None:
            if not self._iterator.hasNext():
                raise StopIteration()
            self._peeked_value = self._iterator.next()
            
        return self._peeked_value

    def next(self):
        """
        :rtype: int
        """
        if self._peeked_value != None:
            ret = self._peeked_value 
            self._peeked_value = None
            return ret  
        if not self._iterator.hasNext():
            raise StopIteration()
        
        return self._iterator.next()

    def hasNext(self):
        """
        :rtype: bool
        """
        return self._peeked_value != None or self._iterator.hasNext()

# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].

File: LeetCode/Problems/C++/284. Peeking Iterator.cpp
/*
 * Below is the interface for Iterator, which is already defined for you.
 * **DO NOT** modify the interface for Iterator.
 *
 *  class Iterator {
 *		struct Data;
 * 		Data* data;
 *		Iterator(const vector<int>& nums);
 * 		Iterator(const Iterator& iter);
 *
 * 		// Returns the next element in the iteration.
 *		int next();
 *
 *		// Returns true if the iteration has more elements.
 *		bool hasNext() const;
 *	};
 */

class PeekingIterator : public Iterator {
public:
    int peekedNext;
	PeekingIterator(const vector<int>& nums) : Iterator(nums) {
	    // Initialize any member here.
	    // **DO NOT** save a copy of nums and manipulate it directly.
	    // You should only use the Iterator interface methods.
	    peekedNext = INT_MIN;
        ;
	}
	
    // Returns the next element in the iteration without advancing the iterator.
	int peek() {
        if(peekedNext!= INT_MIN)
            return peekedNext;
        else{
            peekedNext = Iterator::next();
            return peekedNext;
        }
	}
	
	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	int next() {
	    if(peekedNext!= INT_MIN){
            int ret = peekedNext;
            peekedNext = INT_MIN;
            return ret;
        }
        if(hasNext()){
            return Iterator::next();
        }
        throw "Stop iterator!";

            
	}
	
	bool hasNext() const {
	    return peekedNext!= INT_MIN or Iterator::hasNext();
	}
};